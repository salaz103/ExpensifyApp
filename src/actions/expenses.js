import {v1 as uuid} from 'uuid';

//ADD_EXPENSE
export const addExpense=(
    {
        //AQUI ESTAMOS PONIENDOLE VALORES POR DEFAULT POR SI NO LOS TRAJERA
        description= '', 
        note='',
        amount=0,
        createdAt=0
    }={} //AQUI SE ESTA MANDANDO UN OBJETO VACIO, EN CASO DE QUE EL OBJETO NO EXISTIERA
    )=>({
    type:'ADD_EXPENSE',
    expense:{
        ///LA FUNCION uuid ES PARA GENERAR UN ID UNICO
        id: uuid(),
        description,
        note,
        amount,
        createdAt
    }
});

//REMOVE_EXPENSE
export const removeExpense=({id}={})=>({
    type:'REMOVE_EXPENSE',
    id
});

//EDIT_EXPENSE
export const editExpense=(id,updates)=>({
    type:'EDIT_EXPENSE',
    id,
    updates
});

