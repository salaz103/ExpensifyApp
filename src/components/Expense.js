import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

            //AQUI SE ESTA HACIENDO UN SPREAD (...) DEL OBJETO, PERO TAMBIEN SE PUEDE LLAMAR A "DISPATCH"
const ExpenseListItem= ({description,amount,createdAt,id})=>(
    <Link className="list-item" to={`/edit/${id}`}>
    <div>
    <h3 className="list-item__title">{description}</h3>
    <span className="list-item__subtitle">Created At: {moment(createdAt).format('MMMM Do, YYYY')}</span>
    </div>
    <h3 className="list-item__data">Amount: {numeral(amount/100).format('$0,0.00')}</h3>
    </Link>
);

export default ExpenseListItem;