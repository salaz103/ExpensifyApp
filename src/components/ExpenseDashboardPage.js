import React from 'react';
import ExpenseList from './ExpenseList';


const ExpenseDashboardPage=()=>(
    <ExpenseList/>
);


export default ExpenseDashboardPage;