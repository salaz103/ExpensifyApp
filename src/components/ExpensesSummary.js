import React from 'react';
import { connect } from 'react-redux';
import {Link} from 'react-router-dom';
import selectTotalExpenses from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';
import numeral from 'numeral';



export const ExpensesSummary = (props) => (
    <div className="page-header">
        <div className="content-container">
            <h1 className="page-header__title">Viewing <span>{props.expenseCount}</span>
                {props.expenseCount == 0 || props.expenseCount > 1 ? ' expenses ' : ' expense '}
            totalling <span>{numeral(props.expenseTotal / 100).format('$0,0.00')}</span>
            </h1>
            <div className="page-header__actions">
                <Link className="button" to="/create">Add Expense</Link>
            </div>
        </div>
    </div>
);


const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expenseCount: visibleExpenses.length,
        expenseTotal: selectTotalExpenses(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);