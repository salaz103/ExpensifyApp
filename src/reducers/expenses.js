const expensesReducerDefaultState=[];

const expensesReducer=(state=expensesReducerDefaultState,action)=>{
    switch(action.type){
        case 'ADD_EXPENSE':
            //SE USO "concat" YA QUE ESTE REGRESA UN NUEVO ARREGLO
            // POR QUE SI SE HUBIERA USADO PUSH SE HUBIERA MODIFICADO
            //EL ESTADO, Y EL ESTADO NO DEBE CAMBIARSE
           //return state.concat(action.expense);
           //LOS 3 PUNTOS ... ES UNA FORMA DE EXPANDIR UN ARREGLO
           //ES UN OPERADOR DE EXPASION
           //ES IMPORTANTE RECORDAR QUE NO ESTAMOS CAMBIANDO EL STATE
           //SOLO ESTAMOS REGRESANDO UN ARREGLO PERO LO HACEMOS AL LEER EL "STATE"
           return [...state,action.expense]
        case 'REMOVE_EXPENSE':
            //filter NO CAMBIA EL ARREGLO, ESTE DEVUELVE UNO NUEVO, POR ESO SE UTILIZA
            //AQUI TAMBIEN SE HIZO UN "DESTRUCTURING" PARA RECIBIR EL ID EN LA FUNCION 
            //DEL FILTER
            return state.filter(({id})=> id !== action.id);
        case 'EDIT_EXPENSE':
            //MAP ES OTRA FUNCION QUE REGRESA UN NUEVO ARREGLO
            return state.map((expense)=>{
                if(expense.id===action.id){
                    //SI HAY UN GASTO IGUAL AL QUE QUEREMOS CAMBIAR, SE REGRESA UN NUEVO OBJETO
                    //AQUI UTILIZAMOS EL OPERADOR DE EXPANSION "..." PARA OBJETOS
                    return{
                        ...expense,
                        ...action.updates
                    }
                }else{
                    return expense;
                }
            })
        default:
            return state;
    }
};

export default expensesReducer;