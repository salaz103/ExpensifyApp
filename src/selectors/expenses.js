//GET VISIBLE EXPENSES
//EL 2DO PARAMETRO (FILTERS) ESTA SIENDO "DESTRUCTURING"
const getVisibleExpenses=(expenses,{ text,sortBy, startDate, endDate})=>{
    return expenses.filter((expense)=>{
        
        //EL PRIMERO SIEMPRE VA A DAR TRUE PARA "startDate" o "endDate" QUE NO SEAN NUMEROS,
        //SI DA TRUE EN LA PRIMERA CONDICION, DE IGUAL FORMA FILTRA EL OBJETO (EXPENSE)
        //SI EL "startDate" o "endDate" es undefined LO QUE VA A PASA ES QUE PRACTICAMENTE 
        //NO SE ESTA PONIENDO FILTROS Y POR ENDE ES QUE DA "true" Y CADA OBJETO ES OBTENIDO
        //POR QUE HAY QUE RECORDAR QUE ESTAMOS DENTRO DE LA FUNCION "filter" QUE REGRESA CADA OBJETO
        //SI SE RETORNA "true"
    ///--------------------------------------------------------------------------///
        //SI EL "startDate" o "endDate" ES UN NUMERO, EN LA PRIMERA CONDICION VA A DAR "false"
        //ENTONCES PASA A LA SEGUNDA Y EVALUA, SI EL GASTO FUE CREADO EN UNA FECHA MAYOR O IGUAL 
        //DEL FILTRO QUE SE ESTA PONIENDO "startDate" ENTONCES SE REGRESA "true"
        // Y SI EL GASTO FUE CREADO EN UNA FECHA MENOR O IGUAL QUE EL FILTRO QUE SE ESTA PONIENDO "endDate"
        //ENTONCES SE REGRESA "true".
        const startDateMatch= typeof startDate!=='number' || expense.createdAt>=startDate;
        const endDateMatch= typeof endDate!=='number' || expense.createdAt<=endDate;
        
        const st1=expense.description.toLowerCase();
        const st2= text.toLowerCase();
        const textMatch= st1.includes(st2);

        //POR ULTIMO, SI TODO DIO TRUE, SE RETORNA EL OBJETO QUE LA FUNCION "filter" ESTA LEYENDO
        //DEBIDO A QUE LA LOGICA DE TODO YA SE HIZO ARRIBA
        return startDateMatch && endDateMatch && textMatch;

        //ESTA FUNCION "sort" TIENE UNA FUNCION DE COMPARACION
    }).sort((a,b)=>{
        //SI SE ESTA ORDENANDO POR FECHA
        if(sortBy==='date'){
            //SI "a" TIENE QUE VENIR ANTES, SE PONE -1, PERO SI "b" tiene que venir antes se regresa un 1
            //AQUI LO QUE SE ESTA HACIENDO ES QUE SE ESTA ORDENANDO DESCENDENTEMENTE, ES DECIR CON LA FECHA
            //DE CREACION M√ÅS ANTIGUA, POR EJEMPLO:
            //SI LA CREACION DE "a" ES ANTES QUE LA CREACION DE "b", SE REGRESA UN 1, ES DECIR,
            //SE PONE A "b" PRIMERO, EN OTRAS PALABRAS, DE MAS RECIENTE A MAS ANTIGUA (descendente)
            return a.createdAt< b.createdAt? 1 :-1;
        }else if(sortBy==='amount'){
            ///ESTE TAMBIEN SE VA ORDENAR DE FORMA DESCENDENTE
            return a.amount < b.amount ? 1 :- 1;
        }
    });
};

export default getVisibleExpenses;