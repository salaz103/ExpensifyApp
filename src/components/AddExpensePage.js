import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import {addExpense} from '../actions/expenses';

const AddExpensePage=(props)=>(
    <div>
        <h1>Add Expense</h1>
        <ExpenseForm
            onSubmit={(expense)=>{
                props.dispatch(addExpense(expense));
                //ESTO USA EL BROWSER ROUTING, PARA REDIRECCIONAR A LA PAGINA DEL "DASHBOARD"
                //DESPUES DE HABER DESPACHADO LA ACCION A LA STORE
                props.history.push('/');
            }}
        />
    </div>
);

export default connect()(AddExpensePage);