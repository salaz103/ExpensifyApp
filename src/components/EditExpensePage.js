import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';





export class EditExpensePage extends React.Component {

    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }

    onRemove = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push('/')
    }

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                {/* Editing the expense with ID of {props.match.params.id} */}
                <div className="content-container">
                    <ExpenseForm
                        //ESTE "EXPENSE" ES EL QUE ESTAMOS ENVIANDO PARA QUE LLENE LOS CAMPOS EN EL "FORM"    
                        expense={this.props.expense}
                        //ESTE "EXPENSE" ES EL QUE ESTA SUBIENDO, YA SEA CON INFORMACION NUEVA O SIN CAMBIOS 
                        onSubmit={this.onSubmit}
                    />
                    <button className="button button--secondary" onClick={this.onRemove}>Remove Expense</button>
                </div>
            </div>
        );
    }

}

const mapDispatchToProps = (dispatch) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data))

});


const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id;
        })
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);