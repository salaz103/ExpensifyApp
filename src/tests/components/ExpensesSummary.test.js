import React from 'react';
import {shallow} from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should render Expense summary with various expenses',()=>{
    const wrapper= shallow(<ExpensesSummary expenseCount={expenses} expenseTotal={1200}/>)
    expect(wrapper).toMatchSnapshot();
});

test('should render Expense summary with one expense',()=>{
    const wrapper= shallow(<ExpensesSummary expenseCount={[expenses[0]]} expenseTotal={235}/>)
    expect(wrapper).toMatchSnapshot();
});