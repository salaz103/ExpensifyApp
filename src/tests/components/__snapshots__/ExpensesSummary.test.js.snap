// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expense summary with one expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        <Component />
      </span>
       expense 
      totalling 
      <span>
        $2.35
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render Expense summary with various expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        <Component />
        <Component />
        <Component />
      </span>
       expense 
      totalling 
      <span>
        $12.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
